minidump/src/minidump.rs

```diff
@@ -1084,11 +1089,14 @@ impl Module for MinidumpUnloadedModule {
     fn code_file(&self) -> Cow<'_, str> {
         Cow::Borrowed(&self.name)
     }
-    fn code_identifier(&self) -> CodeId {
-        CodeId::new(format!(
+    fn code_identifier(&self) -> Option<CodeId> {
+        // TODO: This should be returning None if the unloaded module is coming
+        // from a non-Windows minidump. We'll need info about the operating
+        // system, ideally sourced from the SystemInfo to be able to do this.
```

Comment:
IIRC when I looked at unloaded modules I've only seen them from windows' minidump writer.  But I guess that's fairly academic as eventually other writers might start doing this too.

However more generally I'm not sure if this comment is really true?  If say Linux was able to give us an unloaded module it could still have been a really shared object and thus have had a real CodeId I think.

Reply:
Could you clarify the bit about the comment being incorrect? I could see the argument for its inaccuracy in the sense that `MinidumpUnloadedModule::code_identifier` currently would only ever be invoked for Windows minidumps right now, so there wouldn't exist any code paths that would lead to this when parsing non-Windows minidumps. Is that what the second paragraph is trying to convey?

Were you pointing out its inaccuracy with the hopes that the comment would be removed, or simply corrected? Do you feel that commenting on this adds nothing and should be removed?

Reply:
I don't understand the claim that this should always be None on non-Windows, to the extend that having such a comment seems misleading to me.  The 2nd paragraph is trying to explore if there is a reason why a hypothetical linux minidump writer might never be able to produce something that has a CodeId.  And why such a hypothetical writer may need us to have knowledge of the OS to be able to give the correct return value.

I was pointing this out in the hopes that either I am shown to be wrong and the comment could be tweaked to clarify this better, or I'm not wrong and I guess in that case maybe removing the comment rather than tweaking it makes most sense?

Reply:
I see. Unless I've missed something, there simply isn't any way to extract a code ID from a non-Windows minidump. It's just not implemented. 

However, the current implementation still returns an incorrect value for non-Windows minidumps, [if you compare it to breakpad's.](https://chromium.googlesource.com/breakpad/breakpad/+/42d24756174947921d282ee87e2b8e48f6eed3bf/src/processor/minidump.cc#3730) Even without the comparison to breakpad, I think the simplicity of this implementation compared to the one for a loaded module raises at least one red flag about its correctness. Based on breakpad, `code_identifier` for unloaded modules should return `None` unless writers for other non-Windows OSes are changed to write unloaded modules to their minidumps. This makes the first sentence still correct, although it omits the entire story.

If you look at `MINIDUMP_UNLOADED_MODULE` it doesn't look like it stores a CodeView record, meaning that we can't use the strategy used in loaded modules to determine which operating system the unloaded module comes from purely based off of its own contents. I'm sure with some creativity we could use repurpose or augment one of its fields to get this information, but this means that the second sentence in the comment is still correct as well: we don't have enough information in just the unloaded modules stream to fix `code_identifier` so it matches breakpad.

Did this clarify things? I'm under the impression that the comments just need a little bit of rephrasing. It seems like it might be ambiguous about what it is referencing in the second sentence, for example: "why such a hypothetical writer may need us to have knowledge of the OS to be able to give the correct return value" suggests that it gives off the impression that it's talking about the minidump writer instead of the `code_identifier` impl. I do think that removing this TODO would be simply ignoring a spot where the code is (admittedly theoretically) incorrect.

Reply:
> I see. Unless I've missed something, there simply isn't any way to extract a code ID from a non-Windows minidump. It's just not implemented.

Ah, I thought that the CodeId is identical to the DebugId for non-windows as e.g. also described here: https://docs.rs/debugid/latest/debugid/struct.CodeId.html
So I guess that's where our confusion comes from.


But you're right, `MINIDUMP_UNLOADED_MODULE` as defined simply does not have sufficient info here.  I'd be a little surprised if a hypothetical writer on another platform tries to re-use this stream based on that I guess.  But who knows.  Anyway, doesn't do any harm pointing out the breakpad logic.